open_optical_gating/cli/determine_reference_period.py

```diff
@@ -15,50 +15,70 @@
 
 
 def establish(sequence, settings, padding=True):
-    """Establish a reference period from a sequence.
-    Do the processing required for the initial establishing of the heart period
-    (and associated selection of one period's worth of reference frames).
-    Establish the period for the cycle leading up to this current frame.
+    # TODO: JT writes: here and elsewhere, why does this return settings back again?
+    # I’m 99% sure the original settings object will be modified, so returning a new object seems confusing to me.
+    # I can see that returning it could be a reminder that it is changed by the function, but equally it implies to me that
+    # if the caller saved the return value into a *different* variable, the *original* object would be unmodified (which is not the case).
+    # (We can't resolve this by doing a deep copy(), because the object is a large one that contains frame data.
+    #  I would be wary of a shallow copy, because that's just storing up confusion for the future).
+    # I will have a think and try and come up with a solution I like for this general issue.
+    # -> UPDATE: actually, this is bonkers. As far as I can see, all that is updated is the reference period
+    """ Attempt to establish a reference period from a sequence of recently-received frames.
```

Comment:
Ah ok, my mistake. I had just assumed the frame data was in "pog_settings" along with other related state, such as the [float] period, that it naturally belongs with.

What's the best way forward, here? One option would be for me to just go ahead and refactor into a SyncAnalyzer object, at which point I think all this will naturally get taken care of?

Reply:
So, just to be clear, what is the benefit, in your eyes, to using a class rather than a dict (as currently done)? I think I have one benefit in mind but just want to see what you're thinking?

Also, how are we thinking settings should be split? The Settings object should have all the parameters required for the uber OpticalGater class, and then the PiOpticalGater (which inherits from that) would also need a PiSettings that would inherit from Settings but include the extra pi-specific settings needed? So basically, each instance of optical gating would have two objects - an OpticalGater server and a Settings object, both specific to the same of one of file, pi, cam, web?

Reply:
"So, just to be clear, what is the benefit, in your eyes" - I think it's not so much that there is a vast benefit, so much as that (to me at least) the [pog] settings dictionary *is* the state variables for the synchronization process, and so it makes sense to encapsulate that into a class. Just feels more logical and tidy to me. A few concrete benefits though:
- A class can be written to transparently handle the issue of targetSyncPhase being a derived variable.
- Easier to enforce (and self-document) the fact that some of the settings (e.g. numExtraRefFrames) are really constants - certainly not things that should be changing on the fly
- Saves explicit passing around of the dictionary as the first parameter for every function call (and having to return the modified dictionary and/or have ambiguity over whether the function has changed the contents, and indeed who has any right to change the contents).

Reply:
I think I might have misread your comment... I was just about to say 'look I have now refactor OpticalGater as discussed' (caveat - I want to go through and document all three classes a bit better but I need dinner now) shouldn't we just bring all  the settings into the classes and self.attributes... and I now think that was what you were suggesting but I think OpticalGater is SyncAnalyzer - is this what you think too?

Reply:
And we crossed posts there. I think you second reply shows how we've been talking a little at cross purposes. You keep saying incorporate settings into *a* class... but you mean *the* class right? Incorporating all settings as validatable attributes inside OpticalGater, right?

Reply:
"Also, how are we thinking settings should be split?" - I haven't gone through it item by item, but I'm pretty sure there are four categories of "setting"
- Internal state of the synchronization algorithms (e.g. current drift value, reference period, "frameToUseArray"), which I would encapsulate in Sync Analyzer, though it's reasonable for clients to query their values [in most cases].
- Parameters controlling the behaviour of the synchonization algorithms. Storing these in a dictionary is fair enough. Ideally these should all be variables that the client could safely change at any time they feel like it - and they should feel free to do that through the API, just passing a new dictionary for the Sync Analyzer to use.
- Constants [possibly just numExtraRefFrames] controlling the behaviour of the synchronization algorithms, which TBH should probably just be set in __init__ because I can't imagine we'd every actually want to change its value - and certainly shouldn't be changing it on the fly.
- Things that have nothing to do with the sync (e.g. which pins to use for trigger). These would live in a dictionary stored in the [Pi]OpticalGater (and I imagine that sync_settings would be one property within that dictionary). IMO there would be nowt wrong with having a single settings dictionary, and the PiOpticalGater just inserts some extra pi-specific parameters in there. But I haven't thought about that side of things in detail, I've been more interested in the Sync Analyzer side.

Reply:
Again, I think this might require a little further talk in person... I think OpticalGater and SyncAnalyzer are the same thing - but I'm not sure you do... are you thinking SyncAnalyzer is a wrapper around OpticalGater that's just providing the network-protocol?

Reply:
Sorry, this interface is not going well for me. Don't know what is causing all the problems. I'm pretty sure your new comments are not appearing unless I force-refresh the page, so it's not working well as a live conversation!

In terms of terminology (which it now looks like I haven't defined anywhere... in which case: sorry, I could have sworn I did!): in my ObjC code the roles are:
Sync Analyzer encapsulates all the actual sync algorithms etc
Sync Controller handles the GUI *and* [rightly or wrongly] it is indirectly responsible for the camera interfacing etc. So basically, if we are going to separate out the sync algorithms and internal state from "everything else", Sync Controller is "everything else".

Reply:
OK, I've read through your refactor commit. I think what you have done looks good as a start. As per my previous comment, I am proposing to split things out between two independent classes.

One is what I call the Sync Analyzer. This contains all the logic and state for the synchronization algorithms, and which exposes an API in which the caller passes frames or (optionally) an updated parameter dictionary controlling how the sync behaves. It does not know, or care, what platform it is running on, what hardware is connected, whether or not it is running live, etc. _Some_ of the code in OpticalGater will move into that new class. 

The other class is what I call the Sync Controller. This contains "everything else". I see your base class OpticalGater and subclass PiOpticalGater as being equivalent to Sync Controller (as opposed to Sync Analyzer), since it is the Sync Analyzer class hierarchy that knows about what platform it is running on, talks to hardware etc

As I see it, the way to support the M^2 scenario will be for me to write another OpticalGater subclass with the working name SocketOpticalGater, which is pretty much just a stub that listens for messages and forwards them to/from the Sync Analyzer.


Does that arrangement sound reasonable to you, or does that differ from what you were imagining? Happy to discuss.

If you're happy in principle, we should also agree on class names. I'm not convinced that OpticalGater is the most informative name for the bit that _doesn't actually implement the optical gating algorithms_. But equally I'm not fixated on Analyzer/Controller, those are just the names I've got used to myself.

Reply:
I think I disagree slightly. I see OpticalGater as what you're calling SyncAnalyzer. My plans for OpticalGater where to merge setup_camera and load_data into a single 'source' method that is essentially empty and to modify trigger_fluorescence_image_capture to be the trigger out API. In it's current form I don't think OpticalGater does know or care what the hardware is and it is doing all the gating logic.

I then see there being multiple 'Sync Controllers': PiOpticalGater, FileOpticalGater, SocketOpticalGater and VideoOpticalGater that each overwrite the 'source' method with a method that is now aware of the hardware delivering the frames and, optionally, overwrite the trigger_fluorescence_image method to, e.g., send a signal to RPi pins.

Re names - I don't care if we don't call it OpticalGater.... I do care that we should not call it SyncAnalyzer and SyncController - I found it a weird choice of name three years ago and I still agree and my because is: it's not analysing any syncs,  so I don't think it's obvious that SyncAnalyzer controls the sync and SyncContoller uses the sync to control the hardware, and also... the paper titles don't mention sync (I think).